src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,784 @@
+            footprint.max.y = truncate<int>(fast_floor(dy + m_renderer_filter->get_yradius()));
+
+            // Don't affect pixels outside the crop window.
+            footprint = AABB2i::intersect(footprint, crop_window);
+
+            // Bail out if the point does not fall inside the crop window.
+            // Only check the x coordinate; the y coordinate is checked in the loop below.
+            if (footprint.min.x > footprint.max.x)
+                return;
+
+            for (int ry = footprint.min.y; ry <= footprint.max.y; ++ry)
+            {
+                for (int rx = footprint.min.x; rx <= footprint.max.x; ++rx)
+                {
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)][m3hash] += weight; // possible race conditions writing to pixels on the tile edges
+                }
+            }
+            return;
+        }
+    };
+
+    const char* CryptomatteObjectAOVModel = "cryptomatte_object_aov";
+    const char* CryptomatteMaterialAOVModel = "cryptomatte_material_aov";
+}
+
+
+//
+// Cryptomatte AOV class implementation.
+//
+
+struct CryptomatteAOV::Impl
+{
+    vector<WeightMap>                   m_pixel_samples;
+    map<pair<int, int>, NameMap>        m_tile_name_map;
+    unique_ptr<Image>                   m_image;
+    int                                 m_num_layers;
+    CryptomatteAOV::CryptomatteType     m_layer_type;
+};
+
+CryptomatteAOV::CryptomatteAOV(const ParamArray& params)
+  : AOV("cryptomatte", params)
+  , impl(new Impl())
+{
+    const string cryptomatte_type = params.get_optional<string>("cryptomatte_type", "object_names");
+    
+    if (cryptomatte_type == "object_names")
+    {
+        impl->m_layer_type = CryptomatteType::ObjectNames;
+        set_name(CryptomatteObjectAOVModel);
+    }
+    else if (cryptomatte_type == "material_names")
+    {
+        impl->m_layer_type = CryptomatteType::MaterialNames;
+        set_name(CryptomatteMaterialAOVModel);
+    }
+
+    impl->m_num_layers = params.get_optional<int>("cryptomatte_num_layers", 6);
+}
+
+CryptomatteAOV::~CryptomatteAOV()
+{
+    delete impl;
+}
+
+const char* CryptomatteAOV::get_model() const
+{
+    if (impl->m_layer_type == CryptomatteAOV::CryptomatteType::ObjectNames)
+        return CryptomatteObjectAOVModel;
+    else
+        return CryptomatteMaterialAOVModel;
+}
+
+size_t CryptomatteAOV::get_channel_count() const
+{
+    return 0;
+}
+
+const char** CryptomatteAOV::get_channel_names() const
+{
+    return nullptr;
+}
+
+bool CryptomatteAOV::has_color_data() const
+{
+    return false;
+}
+
+void CryptomatteAOV::create_image(
+    const size_t    canvas_width,
+    const size_t    canvas_height,
+    const size_t    tile_width,
+    const size_t    tile_height,
+    ImageStack&     aov_images)
+{
+    const int w = static_cast<int>(canvas_width);
```

Comment:
Why the casts? Those variables don't seem to be useful, just use `canvas_width` and `canvas_height`.